### PR TITLE
[core] Add streaming read from option (#778)

### PR DESCRIPTION
**Pull request by Alibaba-HZY**
Purpose
The user can determine the reading mode of the external log table by configuring the streaming-read-mode option.  If 'streaming-read-mode' = 'log', data will be read from the message queue(log.syetem must be set). If 'streaming-read-mode' = 'file', data will be read from the file store.The  option streaming-read-mode's default value is file when log.syetem is not configue ,and when it is set the default value is log
#778

Tests
LogSystemITCase

API and Format
CoreOptions:

New: streaming-read-mode

Documentation

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 1]
```diff
@@ -559,6 +559,23 @@ public class CoreOptions implements Serializable {
                     .withDescription(
                             "Full compaction will be constantly triggered after delta commits.");
 
+    public static final ConfigOption<StreamReadType> STREAMING_READ_FROM =
+            key("streaming-read-from")
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:563]: `log.read-from`?

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 2]
```diff
@@ -559,6 +559,23 @@ public class CoreOptions implements Serializable {
...
+                    .enumType(StreamReadType.class)
+                    .defaultValue(StreamReadType.FILE_STORE)
```
[the remark this answers is not in the document]
Comment (Alibaba-HZY) [reply paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:565]: ok

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java (hunk 1 above)
Comment (Alibaba-HZY) [reply paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:563]: ok

**Comment by JingsongLi** [comment]
See paimon-docs/README.md.

**Comment by Alibaba-HZY** [comment]
> See paimon-docs/README.md.

I have executed  mvn package -Pgenerate-docs -pl paimon-docs -nsu -DskipTests

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 3]
```diff
@@ -568,6 +568,23 @@ public class CoreOptions implements Serializable {
...
+                            Description.builder()
+                                    .text("Read the store type of the paimon table.")
+                                    .linebreak()
+                                    .linebreak()
+                                    .text("Possible values:")
+                                    .linebreak()
+                                    .list(text("\"file\": Will read from the file store"))
+                                    .list(
+                                            text(
+                                                    "\"log-system\":Will read from the file store, the user must configure log.system"))
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:585]:
```suggestion
                                                    "\"log-system\":Will read from the log store"))
```

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 4]
```diff
@@ -568,6 +568,23 @@ public class CoreOptions implements Serializable {
...
+                    .enumType(StreamReadType.class)
+                    .noDefaultValue()
+                    .withDescription(
+                            Description.builder()
+                                    .text("Read the store type of the paimon table.")
+                                    .linebreak()
+                                    .linebreak()
+                                    .text("Possible values:")
+                                    .linebreak()
+                                    .list(text("\"file\": Will read from the file store"))
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:582]: Use the `StreamReadType.FILE.getValue()`.

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java (hunk 3 above)
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:585]: Ditto.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FlinkSourceBuilder.java [hunk 5]
```diff
@@ -142,7 +144,10 @@ private ContinuousFileStoreSource buildContinuousFileSource() {
 
             // TODO visit all options through CoreOptions
             StartupMode startupMode = CoreOptions.startupMode(conf);
-            if (logSourceProvider == null) {
+            StreamReadType streamReadType = CoreOptions.streamReadType(conf);
+
+            // no log system or
+            if (logSourceProvider == null || streamReadType == FILE) {
```
Comment (SteNicholas) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FlinkSourceBuilder.java:150]: When `logSourceProvider` isn't null and `streamReadType` is `FILE`, is it right to build continuous file source like `latest` startup mode?

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/LogSystemITCase.java [hunk 6]
```diff
@@ -67,4 +67,99 @@ public void testAppendOnlyWithEventual() throws Exception {
...
+                        + "    'fields.word.length' = '1'\n"
+                        + ");");
+        TableResult write =
+                tEnv.executeSql(
+                        "INSERT INTO kafka_file_double_sink SELECT word, COUNT(*) FROM word_table GROUP BY word;");
+        BlockingIterator<Row, Row> read =
+                BlockingIterator.of(
+                        tEnv.executeSql("SELECT * FROM kafka_file_double_sink").collect());
+        List<Row> collect = read.collect(10);
+        assertThat(collect).hasSize(10);
```
Comment (SteNicholas) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/LogSystemITCase.java:111]: Could this verify the actual data content?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FlinkSourceBuilder.java (hunk 5 above)
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FlinkSourceBuilder.java:150]: no log system or log system is configured but streamReadType is FILE

Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FlinkSourceBuilder.java:150]: no log system or log system is configured but streamReadType is FILE

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 7]
```diff
@@ -568,6 +568,27 @@ public class CoreOptions implements Serializable {
                             "Only used to force TableScan to construct 'ContinuousCompactorStartingScanner' and "
                                     + "'ContinuousCompactorFollowUpScanner' for dedicated streaming compaction job.");
 
+    public static final ConfigOption<StreamReadType> STREAMING_READ_FROM =
+            key("streaming-read-from")
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:572]: Could we use `streaming-read-mode` and the value of this option could be `FILE` and `LOG`?

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 8]
```diff
@@ -1054,6 +1079,49 @@ public static FileFormatType fromValue(String value) {
         }
     }
 
+    /** Specifies the type for streaming read. */
+    public enum StreamReadType implements DescribedEnum {
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:1083]:
```suggestion
    public enum StreamingReadMode implements DescribedEnum {
```

**Review by SteNicholas — CHANGES_REQUESTED** [review]
Thanks for @Alibaba-HZY updates. I have left the minor comments. PTAL.

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java (hunk 7 above)
Comment (Alibaba-HZY) [reply paimon-core/src/main/java/org/apache/paimon/CoreOptions.java]: have been modified

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 9]
```diff
@@ -568,6 +568,27 @@ public class CoreOptions implements Serializable {
                             "Only used to force TableScan to construct 'ContinuousCompactorStartingScanner' and "
                                     + "'ContinuousCompactorFollowUpScanner' for dedicated streaming compaction job.");
 
+    public static final ConfigOption<StreamingReadMode> STREAMING_READ_MODE =
+            key("streaming-read-mode")
+                    .enumType(StreamingReadMode.class)
+                    .noDefaultValue()
+                    .withDescription(
+                            Description.builder()
+                                    .text("Read the store type of the paimon table.")
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:577]:
```suggestion
                                    .text("The mode of streaming read that specifies to read the data of table File or Log")
```

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 10]
```diff
@@ -568,6 +568,27 @@ public class CoreOptions implements Serializable {
...
+                                    .linebreak()
+                                    .linebreak()
+                                    .text("Possible values:")
+                                    .linebreak()
+                                    .list(
+                                            text(
+                                                    StreamingReadMode.FILE.getValue()
+                                                            + ": Will read from the file store"))
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:585]:
```suggestion
                                                            + ": Reads from the data of table file."))
```

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 11]
```diff
@@ -568,6 +568,27 @@ public class CoreOptions implements Serializable {
...
+                                    .list(
+                                            text(
+                                                    StreamingReadMode.LOG.getValue()
+                                                            + ":Will read from the log store, the user must configure log.system"))
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:589]:
```suggestion
                                                            + ": Read from the data of table log store."))
```

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 12]
```diff
@@ -1054,6 +1079,49 @@ public static FileFormatType fromValue(String value) {
         }
     }
 
+    /** Specifies the type for streaming read. */
+    public enum StreamingReadMode implements DescribedEnum {
+        LOG("log", "Read from log system such as kafka."),
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:1084]:
```suggestion
        LOG("log", "Reads from log store like kafka etc."),
```

File: paimon-core/src/main/java/org/apache/paimon/CoreOptions.java [hunk 13]
```diff
@@ -1054,6 +1079,49 @@ public static FileFormatType fromValue(String value) {
...
+        FILE("file", "Read from file store.");
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/CoreOptions.java:1085]:
```suggestion
        FILE("file", "Reads from the file store.");
```

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java [hunk 14]
```diff
@@ -132,6 +141,10 @@ private static void validateFileStoreContinuous(Options options) {
             throw new ValidationException(
                     "File store continuous reading dose not support eventual consistency mode.");
```
Comment (SteNicholas) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:142]:
```suggestion
                    "File store continuous reading does not support eventual consistency mode.");
```

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java [hunk 15]
```diff
@@ -113,16 +115,23 @@ static Optional<LogStoreTableFactory> createOptionalLogStoreFactory(
...
 
         return Optional.of(discoverLogStoreFactory(classLoader, configOptions.get(LOG_SYSTEM)));
     }
 
     private static void validateFileStoreContinuous(Options options) {
         LogChangelogMode changelogMode = options.get(LOG_CHANGELOG_MODE);
+        StreamingReadMode streamingReadMode = options.get(STREAMING_READ_MODE);
         if (changelogMode == LogChangelogMode.UPSERT) {
             throw new ValidationException(
                     "File store continuous reading dose not support upsert changelog mode.");
```
Comment (SteNicholas) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:137]:
```suggestion
                    "File store continuous reading does not support upsert changelog mode.");
```

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java [hunk 16]
```diff
@@ -132,6 +141,10 @@ private static void validateFileStoreContinuous(Options options) {
             throw new ValidationException(
                     "File store continuous reading dose not support eventual consistency mode.");
         }
+        if (streamingReadMode == StreamingReadMode.LOG) {
+            throw new ValidationException(
```
Comment (SteNicholas) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:145]: Add the test case for this.

**Review by SteNicholas — APPROVED** [review]
LGTM. Thanks for @Alibaba-HZY updates.

**Comment by JingsongLi** [comment]
@Alibaba-HZY Can you open action for your cloned project?

**Comment by JingsongLi** [comment]
Hi @Alibaba-HZY , why did many force pushing? Is this ready to review now?

**Comment by Alibaba-HZY** [comment]
> Hi @Alibaba-HZY , why did many force pushing? Is this ready to review now?

Previously we set streaming-read-mode default value according to whether log.system is set, and then determine whether to read mq or file only according to this parameter, but when the user executes sql with hint, the parameter option is UnmodifiableMap, it cannot be added to the value. So I changed the if logic. If the user does not configure streamingReadMode but the parameter contains log.system  then to read mq, but does not contain it then  read file, if the streamingReadMode is configured, the decision is made based on it.

**Comment by JingsongLi** [comment]
> > Hi @Alibaba-HZY , why did many force pushing? Is this ready to review now?
> 
> Previously we set streaming-read-mode default value according to whether log.system is set, and then determine whether to read mq or file only according to this parameter, but when the user executes sql with hint, the parameter option is UnmodifiableMap, it cannot be added to the value. So I changed the if logic. If the user does not configure streamingReadMode but the parameter contains log.system then to read mq, but does not contain it then read file, if the streamingReadMode is configured, the decision is made based on it.

Thanks for the information, I suggest you directly push a commit to avoid force pushing.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**